CHSH pulses routine compatible to qibolab 0.2

stavros11 opened the pull request:
Following the previous meeting, this reopens the old PR #1049 which porting the CHSH routine to qibolab 0.2, as now we have properly calibrated two-qubit gates, so I could confirm it works on hardware: http://login.qrccluster.com:9000/xg3XnUDKQfCkJ2jsLh2A7w==

As before, most of the code was copied from the 0.1 routine.

@andrea-pasquale I had to change the test protocol to target qubits `[0, 1]`, otherwise I was getting an error which I think was due to the platform having only two qubits. I am not familiar with the `mock` platform used in the tests, so I am not sure if this is due to that.

andrea-pasquale commented:
> @andrea-pasquale I had to change the test protocol to target qubits `[0, 1]`, otherwise I was getting an error which I think was due to the platform having only two qubits. I am not familiar with the `mock` platform used in the tests, so I am not sure if this is due to that.

Thanks @stavros11, indeed `mock` only has two qubits `[0, 1]`, so most likely that could have been the issue.

andrea-pasquale reviewed (commented):
Thanks @stavros11 for converting this protocol to the 0.2 layout.
Although the CHSH with pulses is for sure useful it would be better to have it for circuits as well, in this way we should be able to deploy to platforms that don't have `CZ` as native gate (otherwise we need to write a specific sequence for each possible two-qubit native).

I know that before using just pulses was useful because there were some problems involving the compiler and flux pulses, however if those are now solved (at least this was my understanding after seeing the latest results from @Luca-Ben-Herrmann and @Edoardo-Pedicillo) I think we could just have a protocol with circuits.

Any thoughts on this @alecandido @Edoardo-Pedicillo @Luca-Ben-Herrmann?

Regarding the PR itself the code is mostly fine. One suggestion is that now we are storing the readout_mitigation matrix directly in `calibration.json`. Therefore we could try to retrieve from there and avoid storing it as `.npz` which should simplify some of the code (as suggested down below).

In general I would suggest to wait until we decide which version of the protocol to keep before doing further changes to this PR.

sorewachigauyo commented:
Also, regarding the usage of the mitigation matrix, how do I load it in?
I have run the readout mitigation for the qubit pair I am targeting and updated the platform `calibration.json`

andrea-pasquale commented:
> Also, regarding the usage of the mitigation matrix, how do I load it in?
> I have run the readout mitigation for the qubit pair I am targeting and updated the platform `calibration.json`

There is a parameter `apply_error_mitigation`, if set to `True` the code will run first the protocol to compute the readout mitigation matrix and it will run the chsh protocol later. This is what is available in the current version.
As I was suggesting above we should modify it to use the readout mitigation stored in `calibration.json` which can be computed using the `readout_mitigation_matrix`.

sorewachigauyo commented:
Okay, I tried running this with mitigation=True
```yaml
platform: sinq-20

actions:
- id: chsh_pulses
  operation: chsh_pulses
  targets: [[8, 3]]
  parameters:
    bell_states: [0]
    ntheta: 20
    native: True
    apply_error_mitigation: True
```
but its complaining about the order, so I will try `[[3, 8]]` next

```python
Traceback (most recent call last):
  File "/mnt/home/nqch_paul/Documents/envs/qibo/bin/qq", line 8, in <module>
    sys.exit(command())
             ^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/click/core.py", line 1442, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/click/core.py", line 1363, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/click/core.py", line 1830, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/click/core.py", line 1226, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/click/core.py", line 794, in invoke
    return callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/qibocal/cli/_base.py", line 77, in run
    protocols_execution(runcard, folder, force, update)
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/qibocal/cli/run.py", line 44, in protocols_execution
    history = runcard.run(
              ^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/qibocal/auto/runcard.py", line 62, in run
    instance.run_protocol(
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/qibocal/auto/execute.py", line 127, in run_protocol
    completed = task.run(platform=self.platform, targets=self.targets, mode=mode)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/qibocal/auto/task.py", line 173, in run
    completed.results, completed.results_time = operation.fit(completed.data)
                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/qibocal/auto/operation.py", line 40, in wrapper
    out = func(*args, **kwds)
          ^^^^^^^^^^^^^^^^^^^
  File "/mnt/home/nqch_paul/Documents/envs/qibo/lib/python3.12/site-packages/qibocal/protocols/two_qubit_interaction/chsh/protocol.py", line 322, in _fit
    matrix = data.mitigation_matrix[pair]
             ~~~~~~~~~~~~~~~~~~~~~~^^^^^^
KeyError: (3, 8)
```

sorewachigauyo commented:
[chsh.tar.gz](https://github.com/user-attachments/files/20601864/chsh.tar.gz)

This runs fine for all the Bell states, but on the UI side maybe a subtitle for each Bell state would be clearer.

stavros11 commented:
In bccfe8caff96adee509f446cb242663966529582 I dropped the error mitigation matrix calculation from the CHSH routine, in favor of using the one that is already saved in calibration.json. If the matrix is not available, then the mitigated frequencies calculation and plot will be skipped, as before. I implemented some getters and setters in the `Calibration` class to simplify this, as I noticed that it is saved as a sparse array and the indexing was not very trivial (from the error mitigation routine).

> Since I feel like everybody agreed in the meeting I think that we can just have the circuit version. Perhaps it could be worth to do just one last test to check if there are any discrepancies and then we could drop the pulse sequence version of the protocol from the PR.

Yes, I agree with this. Unfortunately it seems that the two-qubit gate I used to test it on qw5q_platinum two days ago is no longer calibrated so we will need to calibrate to do this test, but I should be able to do this next week the latest.

stavros11 commented:
I ported the circuit versions from 0.1 and removed the pulses version, as agreed. I have not tested on hardware, however I checked the total waveform arrays per channel and confirmed that they agree between the pulses version and the two circuit versions. Therefore, if we notice any significant differences on hardware, these are probably caused by some other part of the software (eg. driver).

Also, I kept the two circuit versions we had in 0.1: `native: True/False`, to generate the circuit with native gates (GPI2, CZ) or no. In the above test the two were equivalent, so we could even keep only one to simplify, but that's also up to discussion.

Other than this and the missing hardware test, in terms of code this should be ready.

andrea-pasquale commented:
> @andrea-pasquale do you know what is the motivation to sort there?

I don't exactly remember why we are doing it. The problem seems to appear also in other protocols where we try to patch this using something like this:
https://github.com/qiboteam/qibocal/blob/2ede5ecbc9721859f50a0527ba84c388e14f7d5e/src/qibocal/protocols/two_qubit_interaction/chevron/chevron.py#L223-L224

There is already an issue open about a similar thing https://github.com/qiboteam/qibocal/issues/699

For the time being I would propose to just write a patch similar to the one above. We can later address this issue for all two qubit gates protocols.

stavros11 commented:
> Thanks @stavros11, overall it looks good to me. Apart from the possible patch above it would be beneficial to add the protocol to the documentation (similarly to what we are doing with the other protocols). Writing just a single page where we show a possible report could be enough. We could also merge as it is an open an issue about adding the documentation so that we don't forget.

Thanks for the review. I added a patch that fixes the qubit ordering issue for the mock platform in 565abe3cb37024809b66e6bd70d84758230167ff and a doc page in 6c6fa393302c04c64b53eb02ae216b8babf68df9. Maybe the description of the CHSH could be expanded a bit, but I am not familiar with all the details of the implementation as I just copied it from 0.1 and I didn't find any documentation there.

stavros11 commented on src/qibocal/protocols/two_qubit_interaction/chsh/circuits.py:
```diff
@@ -0,0 +1,93 @@
+    p = [0, 0]
+    c.add(gates.GPI2(0, np.pi / 2))
+    c.add(gates.GPI2(1, np.pi / 2))
+    c.add(gates.CZ(0, 1))
```
I think this could also be fixed in the transpiler level. For example, now both the native and non-native versions seem to work on hardware. The non-native circuits contain `H` gates which I believe are not native (there is no compiler rule), so it seems that the transpiler is working. I guess if `CZ` is not available, it should transpile to `CNOT`/`iSWAP` but I am not sure if that is happening currently.

andrea-pasquale replied:
This is true, however I think that the native circuit should be executable even without the transpiler. I was thinking that depending on the two qubit gate available we could have a dedicate native circuit for the CHSH inequality.  In this case we could just add one with the `iSWAP` and another one with the `CNOT`. I see this just as an additional feature that we could add later on.

andrea-pasquale commented:
> Thanks for the review. I added a patch that fixes the qubit ordering issue for the mock platform in [565abe3](https://github.com/qiboteam/qibocal/commit/565abe3cb37024809b66e6bd70d84758230167ff) and a doc page in [6c6fa39](https://github.com/qiboteam/qibocal/commit/6c6fa393302c04c64b53eb02ae216b8babf68df9). Maybe the description of the CHSH could be expanded a bit, but I am not familiar with all the details of the implementation as I just copied it from 0.1 and I didn't find any documentation there.

Thanks, for the time being the documentation looks good. Later once we add also Mermin we could revisit it and expand it a bit more.
